feature-libs/product-configurator/rulebased/components/attribute/price-change/configurator-attribute-price-change.service.ts

```diff
@@ -0,0 +1,118 @@
+    ConfiguratorRouterExtractorService
+  );
+  protected configuratorCommonsService = inject(ConfiguratorCommonsService);
+  protected isInitialRendering: boolean = true;
+  protected lastAttributeSupplement:
+    | Configurator.AttributeSupplement
+    | undefined;
+  protected valuePrices: { [key: string]: Configurator.PriceDetails } = {};
+
+  /**
+   * Returns an observable that shall be used by all components supporting delta rendering mode.
+   * It will monitor the price supplements of configuration observable and emit true if price supplements
+   * matching the given attribute key have changed.
+   * Additionally it returns always true for the first emission of the underlying configuration observable.
+   * This ensures that an enclosing UI component will initially render, even if the async pricing request ist still running,
+   * so that the UI is not blocked. Afterwards a rerender shall only occur if prices change.
+   * This all assumes that the enclosing UI component itself gets recreated or rerendered (triggered elsewhere) whenever the attribute itself changes content wise.
+   *
+   * @param attributeKey key of the attribute for which the prices should be checked for changes
+   * @returns observable that emits 'true' each time a price changes and hence there is the need to rerender the enclosing component
+   */
+  getPriceChangedEvents(attributeKey: string | undefined): Observable<boolean> {
+    return this.configuratorRouterExtractorService.extractRouterData().pipe(
+      switchMap((routerData) => {
+        return this.configuratorCommonsService
+          .getConfiguration(routerData.owner)
+          .pipe(
+            // Initially render attribute without prices, so UI is not blocked, otherwise only re-ender if prices changed.
+            // Changes of attribute itself are already handled in the attribute composition directive.
+            filter(
+              (config) => this.isInitialRendering || !!config.priceSupplements
+            ),
+            switchMap((config) => {
+              if (!config.priceSupplements) {
+                return of(true);
+              }
+              const pricesChanged = this.checkForValuePriceChanges(
+                config,
+                attributeKey
+              );
+              return pricesChanged ? of(true) : EMPTY;
+            }),
+            tap(() => (this.isInitialRendering = false))
+          );
+      })
+    );
+  }
+
+  /**
+   * Merges the stored value price data into the given value, if available.
+   * As the value might be read-only a new object will be returned combining price and value.
+   *
+   * @param value the value
+   * @returns the new value with price
+   */
+  mergePriceIntoValue(value: Configurator.Value): Configurator.Value {
+    const valueName = value.name;
+    if (valueName && this.valuePrices[valueName]) {
+      value = { ...value, valuePrice: this.valuePrices[valueName] };
+    }
+    return value;
+  }
```

Comment:
The name of it is a bit misleading. 
- A the first sight I pretty much thought it's a kinda "setter" -> setting the given data in the state of the class. But after reading the implementation it carefully, I realised it's a kinda "getter" and has no side effects.
- it's not clear whether where the price is coming from (from the given Value param?). I understand that not. But it's a rather `currentPrice` that is known only by the service.

For easier reasoning about this method (without needing to check its implementation), what do you think about renaming it to `getWithCurrentPrice(value: Configurator.Value): Configurator.Value`?

Reply:
[MAJOR]

We're mixing a reactive programing (Rxjs observable flow in `getPriceChangedEvents()`) with static properties (accessing data statically via `mergePriceIntoValue` which were stored as a side effect of observable flow of`getPriceChangedEvents()`), with is generally hard to reason about and btw. error prone in async edge cases.

Ideally, I'd suggest that we should rather have no static getter methods like `mergePriceIntoValue` for our reactive data, but instead emit from our returned `Observable` all the the data needed by the consumers e.g.:

```diff
- getPriceChangedEvents(attributeKey: string | undefined): Observable<boolean> {
+ getAttributePrices(attributeKey: string | undefined): Observable<
+   { pricesByValueName: Record<string, Configurator.PriceDetails }
+ > {
    /* ... */
-               return pricesChanged ? of(true) : EMPTY;
+               return pricesChanged ? of({ this.pricesByValueName }) : EMPTY;
```

Then in the consuming component we could have:
```ts
pricesByValueName$: Observable<{ pricesByValueName: Record<string, Configurator.PriceDetails }>
```

```html
<ng-container *ngIf="(attributePrices$ | async).pricesByValueName as pricesByValueName">
   <!-- ... -->
   <option
     [label]="getLabel(expMode, item.valueDisplay, item.valueCode, this.mergePriceIntoValue(item, pricesByValueName[item.name])"
```

What do you think about it?

Reply:
will be covered in: https://github.com/SAP/spartacus/pull/19097